llvm/include/llvm/Telemetry/Telemetry.h

```diff
@@ -0,0 +1,136 @@
+  // telemetry data (Could be stdout, stderr, or some local paths, etc).
+  //
+  // These strings will be interpreted by the vendor's code.
+  // So the users must pick the from their vendor's pre-defined
+  // set of Destinations.
+  std::vector<std::string> AdditionalDestinations;
+};
+
+/// For isa, dyn_cast, etc operations on TelemetryInfo.
+typedef unsigned KindType;
+/// This struct is used by TelemetryInfo to support isa<>, dyn_cast<>
+/// operations.
+/// It is defined as a struct (rather than an enum) because it is
+/// expectend to be extended by subclasses which may have
+/// additional TelemetryInfo types defined to describe different events.
+struct EntryKind {
+  static const KindType Base = 0;
+};
+
+/// TelemetryInfo is the data courier, used to move instrumented data
+/// from the tool being monitored to the Telemery framework.
+///
+/// This base class contains only the basic set of telemetry data.
+/// Downstream implementations can add more fields as needed to describe
+/// additional events.
+///
+/// For example, The LLDB debugger can define a DebugCommandInfo subclass
+/// which has additional fields about the debug-command being instrumented,
+/// such as `CommandArguments` or `CommandName`.
+struct TelemetryInfo {
+  // This represents a unique-id, conventionally corresponding to
+  // a tool's session - i.e., every time the tool starts until it exits.
+  //
+  // Note: a tool could have multiple sessions running at once, in which
+  // case, these shall be multiple sets of TelemetryInfo with multiple unique
+  // ids.
+  //
+  // Different usages can assign different types of IDs to this field.
+  std::string SessionId;
+
+  TelemetryInfo() = default;
+  virtual ~TelemetryInfo() = default;
+
+  virtual json::Object serializeToJson() const;
+
+  // For isa, dyn_cast, etc, operations.
+  virtual KindType getKind() const { return EntryKind::Base; }
+  static bool classof(const TelemetryInfo *T) {
+    if (T == nullptr)
+      return false;
+    return T->getKind() == EntryKind::Base;
+  }
+};
+
+/// This class presents a data sink to which the Telemetry framework
+/// sends data.
+///
+/// Its implementation is transparent to the framework.
+/// It is up to the vendor to decide which pieces of data to forward
+/// and where to forward them.
+class Destination {
+public:
+  virtual ~Destination() = default;
+  virtual Error emitEntry(const TelemetryInfo *Entry) = 0;
+  virtual llvm::StringLiteral name() const = 0;
+};
+
+/// This class is the main interaction point between any LLVM tool
+/// and this framework.
+/// It is responsible for collecting telemetry data from the tool being
+/// monitored and transmitting the data elsewhere.
+class Telemeter {
+public:
+  // Invoked upon tool startup
+  virtual void atStartup(llvm::StringRef ToolPath, TelemetryInfo *Entry) = 0;
+
+  // Invoked upon tool exit.
+  virtual void atExit(llvm::StringRef ToolPath, TelemetryInfo *Entry) = 0;
```

Comment:
I'm not sure if I expressed this in the past, but I still take issue with these two methods, I don't think they belong in the base implementation and should be specific instance of `TelemtryInfo`. This only makes sense in the context of a tool and not in the context of library like LLDB. When used from Python the calls to Initialize and Terminate do not have to correspond to startup or exit.

Reply:
I still think a generic `dispatch` (or alternatively named function) that takes a `TelemetryInfo` and forwards it to the configured `Destination`(s) would make more sense. Then in the tool/library code, it simply calls `dispatch` whenever required (which could include during startup/close down) and passes in the desired `TelemetryInfo` to do what it wants.

As well as the `dispatch` method, I think a `std::vector<std::unique_ptr<Destination>>` probably belongs in the `Manager` class (which `dispatch` uses). Otherwise, where are the `Destination` instances owned? If the downstream vendor has an implementation that needs to own stuff, then the `Destination` stored in `Manager` could just be a thing that references the downstream vendor piece.

`dispatch` would look something like:
```
void dispatch(TelemetryInfo *event) {
  // Maybe have an `onDispatch` hook that is vendor-defined (default do 
  // nothing to allow the vendor to do extra stuff here).
  for (Destination *dest : destinations)
    dest->receiveEntry(event);
}
```
Then in the tool code, you might have something like:
```
int tool_main(...) {
  ...
  manager.dispatch(StartupEvent());
  ...
  manager.dispatch(OtherEvent());
  ...
  manager.dispatch(ShutdownEvent());
}
```
Finally, I think it makes sense for the `Manager` constructor to take a `Config`, so that the downstream implementations of `Manager` can then hook into the stuff defined by `Config` in some manner. In particular, I could see an argument that the `dispatch` method should do nothing when telemetry is not enabled, so it would need access to that information.

Reply:
done - changed this to `dispatch()`